File: app/src/main/java/com/example/advancedprayertimes/Logic/LocationUtil.java
package com.example.advancedprayertimes.Logic;

import android.content.Context;
import android.location.Address;
import android.location.Geocoder;

import com.example.advancedprayertimes.Logic.Entities.CustomLocation;

import java.util.List;
import java.util.Locale;

public class LocationUtil
{
    public static Address RetrieveCityByLocation(Context context, CustomLocation location) throws Exception
    {
        Geocoder geocoder = new Geocoder(context, Locale.getDefault());

        List<Address> addresses = geocoder.getFromLocation(location.getLatitude(),location.getLongitude(),5);

        if (Geocoder.isPresent() && addresses.size() > 0)
        {
            return addresses.get(0);
        }

        return null;
    }

    public static List<Address> RetrieveCitiesByName(Context context, String name) throws Exception
    {
        Geocoder geocoder = new Geocoder(context);
        return geocoder.getFromLocationName(name, 5);
    }
}

File: app/src/main/kotlin/com/example/advancedprayertimes/logic/util/LocationUtil.kt
package com.example.advancedprayertimes.logic.util

import android.content.Context
import android.location.Address
import kotlin.Throws
import android.location.Geocoder
import java.lang.Exception
import java.util.*

object LocationUtil
{
    @JvmStatic
    @Throws(Exception::class)
    fun RetrieveCityByLocation(context: Context?, longitude: Double, latitude: Double): Address?
    {
        val geocoder = Geocoder(context, Locale.getDefault())
        val addresses = geocoder.getFromLocation(latitude, longitude, 5)

        return if (Geocoder.isPresent() && addresses.size > 0)
        {
            addresses[0]
        }
        else null
    }

    @Throws(Exception::class)
    fun RetrieveCitiesByName(context: Context?, name: String?): List<Address>
    {
        val geocoder = Geocoder(context)
        return geocoder.getFromLocationName(name, 5)
    }
}